doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+
+### Overview
+
+For high availability, a system may run multiple replicas of a gNMI client (controller). Among the replicas, only one client should be elected as master and do gNMI operations that mutate any state on the target. However, in the event of a network partition, there can be two or more replicas thinking themselves as master. For read-only RPCs, this is OK. But if they both call the `Set` RPC, the target may be incorrectly configured by the stale master. Therefore, "Master Arbitration" is needed when multiple clients exist.
+
+**Definition**
+
+When Master Arbitration feature is turned on, the gNMI server receives a proto with the election ID (EID) specified in it. The election ID is a 128-bit integer MUST be guaranteed to be monotonically increasing -- a new master's election ID MUST always be larger than its predecessor. The process through which election IDs are allocated to client replicas is out-of-scope of this specification.
+
+The client carries the election ID in all of its `Set` requests. When the target receives any `Set` request, it looks at the election ID carried with the requests and stores the largest ID it sees. An election ID that is equal to or greater than the currently stored one should be accepted as the master. On the other hand, Set requests with no election ID or with a smaller election ID are rejected and return an error.
+
+**Client Role**
+
+Sometimes, there is a need to partition the config tree among multiple clients
+(where each client in turn could be replicated). This is accomplished by
+assigning a *role* to each client. Master Arbitration is performed on a per-role
+basis. There can be one master for each role, i.e. multiple master clients are
+allowed only if they belong to different roles.
+
+The role is identified by its `id`, which is assigned offline in agreement
+across the entire control plane. The target will use the role ID to arbitrate
+such that each role has one and only one master controller instance.
+
+To simplify for use-cases where multiple roles are not needed, the client can
+leave the role message unset in MasterArbitration. This implies default role.
+All `Set` RPCs that uses a default role are arbitrated in the same group.
+
+### Requirements
+
+Master Arbitration is a HA (high-availability) SDN-focused feature therefore it cannot be enabled by default. On the other hand, when it is required, it has to be enabled from the very moment the gNMI server starts. As an optional feature, Master Arbitration could be turned off when gNMI server restarted. And it should not impact the overall perfermance.
+
+### Architecture Design
+
+No architecture change is needed for Master Arbitration. The protobuf changes is done already. The protobuf of a `MasterArbitration` message is defined as following:
+
+```
+message MasterArbitration {
+  Role role = 1;
+  Uint128 election_id = 2;
+}
+
+message Uint128 {
+  uint64 high = 1;
+  uint64 low = 2;
```

Comment:
What is the behaviour in gNMI server when there are changes in the low & high values of the EID in the set Request?

Reply:
The uint128 type comparison is implemented [here](https://github.com/sonic-net/sonic-gnmi/pull/96/files#diff-27de8bb6147404b50ed13655e76620985b572253e2e8105ebad6d2d6bc4e8c0fR536). high values are compared first. If there is a tie in high values, low values are compared.